File: 01-content-factory/integrations/slack.ts
import type { GeneratedContent, QAResult } from "../types";

export interface SlackNotification {
  channel?: string;
  text: string;
  blocks?: SlackBlock[];
}

export interface SlackBlock {
  type: string;
  text?: { type: string; text: string };
  elements?: any[];
  accessory?: any;
}

export async function sendSlackNotification(
  notification: SlackNotification
): Promise<{ success: boolean; error?: string }> {
  const webhookUrl = process.env.SLACK_WEBHOOK_URL;
  
  if (!webhookUrl) {
    console.log("[Slack] Webhook URL not configured - skipping notification");
    return { success: false, error: "SLACK_WEBHOOK_URL not configured" };
  }

  try {
    const response = await fetch(webhookUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(notification),
    });

    if (!response.ok) {
      throw new Error(`Slack API error: ${response.statusText}`);
    }

    return { success: true };
  } catch (error: any) {
    console.error("[Slack] Failed to send notification:", error.message);
    return { success: false, error: error.message };
  }
}

export async function notifyContentReady(
  content: GeneratedContent,
  qaResult: QAResult,
  approvalUrl: string
): Promise<void> {
  const statusEmoji = qaResult.passed ? "✅" : "⚠️";
  const scoreColor = qaResult.score >= 90 ? "good" : qaResult.score >= 75 ? "warning" : "danger";

  const notification: SlackNotification = {
    text: `${statusEmoji} New content ready for review: ${content.title}`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: `${statusEmoji} Content Ready for Review`,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*${content.title}*\n_Type: ${content.type} | Score: ${qaResult.score}/100_`,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Preview:*\n\`\`\`${content.content.substring(0, 300)}${content.content.length > 300 ? "..." : ""}\`\`\``,
        },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: { type: "plain_text", text: "✅ Approve" },
            style: "primary",
            url: `${approvalUrl}?action=approve&id=${content.id}`,
          },
          {
            type: "button",
            text: { type: "plain_text", text: "❌ Reject" },
            style: "danger",
            url: `${approvalUrl}?action=reject&id=${content.id}`,
          },
          {
            type: "button",
            text: { type: "plain_text", text: "👀 View Full" },
            url: `${approvalUrl}?id=${content.id}`,
          },
        ],
      },
    ],
  };

  await sendSlackNotification(notification);
}

export async function notifyRunComplete(
  runId: string,
  stats: {
    totalPieces: number;
    passed: number;
    failed: number;
    byType: Record<string, number>;
  },
  dashboardUrl: string
): Promise<void> {
  const typeBreakdown = Object.entries(stats.byType)
    .map(([type, count]) => `• ${type}: ${count}`)
    .join("\n");

  const notification: SlackNotification = {
    text: `🚀 Content run ${runId} completed! ${stats.totalPieces} pieces generated.`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: "🚀 Content Factory Run Complete",
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Run ID:* \`${runId}\`\n*Total Pieces:* ${stats.totalPieces}\n*Passed QA:* ${stats.passed}\n*Needs Review:* ${stats.failed}`,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Breakdown by Type:*\n${typeBreakdown}`,
        },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: { type: "plain_text", text: "📊 View Dashboard" },
            style: "primary",
            url: dashboardUrl,
          },
        ],
      },
    ],
  };

  await sendSlackNotification(notification);
}

File: 01-content-factory/python/integrations/slack.py
"""
Slack Integration for QA approval notifications
"""
import os
import httpx
from typing import Optional


class SlackNotifier:
    """Slack webhook notifier for content approvals"""
    
    def __init__(self):
        self.webhook_url = os.environ.get("SLACK_WEBHOOK_URL")
        self.available = bool(self.webhook_url)
        
        if not self.available:
            print("SLACK_WEBHOOK_URL not set - Slack notifications disabled")
    
    async def send_message(self, text: str) -> bool:
        """Send a simple text message"""
        if not self.available:
            print(f"[Slack] Mock message: {text}")
            return True
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    self.webhook_url,
                    json={"text": text},
                    timeout=10.0,
                )
                return response.status_code == 200
            except Exception as e:
                print(f"Slack error: {e}")
                return False
    
    async def send_approval_request(
        self,
        content_id: str,
        client_name: str,
        content_type: str,
        title: str,
        preview: str,
        qa_score: float,
        approve_url: str,
        reject_url: str,
    ) -> bool:
        """Send approval request with action buttons"""
        if not self.available:
            print(f"[Slack] Mock approval request: {title} (score: {qa_score})")
            return True
        
        blocks = [
            {
                "type": "header",
                "text": {
                    "type": "plain_text",
                    "text": f"🎯 New Content Ready for {client_name}",
                    "emoji": True
                }
            },
            {
                "type": "section",
                "fields": [
                    {"type": "mrkdwn", "text": f"*Type:*\n{content_type}"},
                    {"type": "mrkdwn", "text": f"*QA Score:*\n{qa_score}/10"},
                ]
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"*{title}*\n\n{preview[:500]}..."
                }
            },
            {
                "type": "actions",
                "elements": [
                    {
                        "type": "button",
                        "text": {"type": "plain_text", "text": "✅ Approve", "emoji": True},
                        "style": "primary",
                        "url": approve_url,
                        "action_id": f"approve_{content_id}"
                    },
                    {
                        "type": "button",
                        "text": {"type": "plain_text", "text": "❌ Reject", "emoji": True},
                        "style": "danger",
                        "url": reject_url,
                        "action_id": f"reject_{content_id}"
                    }
                ]
            }
        ]
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    self.webhook_url,
                    json={"blocks": blocks},
                    timeout=10.0,
                )
                return response.status_code == 200
            except Exception as e:
                print(f"Slack error: {e}")
                return False
    
    async def send_run_summary(
        self,
        client_name: str,
        run_id: str,
        total_pieces: int,
        passed_pieces: int,
        failed_pieces: int,
    ) -> bool:
        """Send summary after a content run completes"""
        
        emoji = "🎉" if passed_pieces == total_pieces else "⚠️"
        
        message = f"""
{emoji} *Content Run Complete for {client_name}*

*Run ID:* `{run_id}`
*Total Pieces:* {total_pieces}
*Passed QA:* {passed_pieces} ✅
*Failed QA:* {failed_pieces} ❌

Review pending content in the dashboard.
"""
        
        return await self.send_message(message)
